File: Data Structures & Algorithms/add-two-numbers/submission-5.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        
        ListNode sum = new ListNode (0);
        ListNode temp = sum;
        int carry = 0;

        while (l1 != null && l2 != null) {
            int digit = l1.val + l2.val + carry;
            carry = digit / 10;
            temp.next = new ListNode(digit % 10);

            l1 = l1.next;
            l2 = l2.next;
            temp = temp.next;
        }

        while (l1 != null) {
            int digit = l1.val + carry;
            carry = digit / 10;
            temp.next = new ListNode (digit % 10);

            l1 = l1.next;
            temp = temp.next;
        }

        while (l2 != null) {
            int digit = l2.val + carry;
            carry = digit / 10;
            temp.next = new ListNode (digit % 10) ;

            l2 = l2.next;
            temp = temp.next;
        }

        if (carry > 0) {
            temp.next = new ListNode (carry);
        }

        return sum.next;
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-5.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if(lists.empty()) return nullptr;

        struct Compare{
            bool operator()(ListNode* a, ListNode* b) {
                return a->val > b->val;
            }
        };

        priority_queue<ListNode*, vector<ListNode*>, Compare> pq;
        for(const auto& list : lists) {
            if(list != nullptr) pq.push(list);
        }

        ListNode* result = new ListNode(0);
        ListNode* dummy = result;

        while(!pq.empty()) {
            ListNode* curr = pq.top();
            pq.pop();

            dummy->next = curr;
            dummy = dummy->next;

            if(curr->next != nullptr) {
                pq.push(curr->next);
            }
        }

        return result->next;
    }
};
